### PR TITLE
implement multi-language slug.

### PR DESCRIPTION
**Pull request by nguyendtu**
multilanguge slug.

### CODE REVIEW

File: app/modules/Page/Model/Page.php [hunk 1]
```diff
@@ -65,9 +66,17 @@ public function validation()
...
+        $slugRecord = Translate\PageTranslate::findFirst([$queryML, 'cache' => array('key' => $key, 'lifetime' => 60)]);
+
         $key = HOST_HASH . md5("Page::findFirst($query)");
-        $page = self::findFirst(array($query, 'cache' => array('key' => $key, 'lifetime' => 60)));
+        $page = self::findFirst([
+          'conditions' => 'id = :pageId:',
+          'bind' => ['pageId' => $slugRecord->foreign_id],
+          'cache' => array('key' => $key, 'lifetime' => 60)
+          ]);
+          var_dump($pages);
```
Comment (sergeyklay) [inline app/modules/Page/Model/Page.php]: Could you please remove this

File: app/modules/Page/Model/Page.php [hunk 2]
```diff
@@ -65,9 +66,17 @@ public function validation()
...
     {
-        $query = "slug = '$slug'";
+        $queryML = "key = 'slug' and value = '$slug'";
+        $key = HOST_HASH . md5("Page::findFirst($queryML)");
+        $slugRecord = Translate\PageTranslate::findFirst([$queryML, 'cache' => array('key' => $key, 'lifetime' => 60)]);
+
         $key = HOST_HASH . md5("Page::findFirst($query)");
-        $page = self::findFirst(array($query, 'cache' => array('key' => $key, 'lifetime' => 60)));
+        $page = self::findFirst([
+          'conditions' => 'id = :pageId:',
```
Comment (sergeyklay) [inline app/modules/Page/Model/Page.php]: Fix indents please

File: app/modules/Page/Model/Page.php [hunk 3]
```diff
@@ -65,9 +66,17 @@ public function validation()
...
+          'bind' => ['pageId' => $slugRecord->foreign_id],
+          'cache' => array('key' => $key, 'lifetime' => 60)
```
Comment (sergeyklay) [inline app/modules/Page/Model/Page.php]: `array('key' => $key, 'lifetime' => 60)` => `['key' => $key, 'lifetime' => 60]`

File: app/modules/Page/Model/Page.php [hunk 4]
```diff
@@ -65,9 +66,17 @@ public function validation()
 
     public static function findCachedBySlug($slug)
     {
-        $query = "slug = '$slug'";
+        $queryML = "key = 'slug' and value = '$slug'";
+        $key = HOST_HASH . md5("Page::findFirst($queryML)");
```
Comment (sergeyklay) [inline app/modules/Page/Model/Page.php]: how about:
```php
$key = HOST_HASH . md5(__METHOD__ . $queryML);
```

File: app/modules/Application/Mvc/Helper/LangSwitcher.php [hunk 5]
```diff
@@ -17,10 +17,9 @@ public function render($lang, $string)
         $router = $this->getDI()->get('router');
         $view = $this->getDI()->get('view');
         $url = $this->getDI()->get('url');
-        $requestQuery = new RequestQuery();
 
+        $requestQuery = new RequestQuery();
```
Comment (sergeyklay) [inline app/modules/Application/Mvc/Helper/LangSwitcher.php]: could you please explain why?

File: app/modules/Application/Mvc/Helper/LangSwitcher.php [hunk 6]
```diff
@@ -31,10 +30,14 @@ public function render($lang, $string)
                 $url_args['for'] = $route_name_changed;
 
                 $route_params = $router->getParams();
+
```
Comment (sergeyklay) [inline app/modules/Application/Mvc/Helper/LangSwitcher.php]: why?

File: app/modules/Application/Mvc/Helper/LangSwitcher.php [hunk 7]
```diff
@@ -70,4 +73,28 @@ private function changeRouteName($route_name, $lang)
         return $route_name . '_' . $lang;
     }
 
-} 
+    private function getMLSlug($matched_route, $route_params, $lang)
```
Comment (sergeyklay) [inline app/modules/Application/Mvc/Helper/LangSwitcher.php]:
```php
use Phalcon\Mvc\Router\Route;

protected function getMLSlug(Route $route, array $params, $lang)
{
    // ...
}
```

File: app/modules/Application/Mvc/Helper/LangSwitcher.php [hunk 8]
```diff
@@ -70,4 +73,28 @@ private function changeRouteName($route_name, $lang)
...
+        $paths = $matched_route->getPaths();
+        $module = $paths['module'];
+
+        $curSLug = $route_params['slug'];
+        // query moudle entity id.
+        $query = "key = 'slug' and value = '$curSLug'";
+        $modelName = \Phalcon\Text::camelize($module);
+        $modelTranslateName = "{$modelName}Translate";
+        $translateModel = "{$modelName}\Model\Translate\\{$modelName}Translate";
+        $slugRecord = $translateModel::findFirst([$query]);
```
Comment (sergeyklay) [inline app/modules/Application/Mvc/Helper/LangSwitcher.php]: In general this code looks like pasta factory

File: app/modules/Application/Mvc/Helper/LangSwitcher.php [hunk 9]
```diff
@@ -70,4 +73,28 @@ private function changeRouteName($route_name, $lang)
...
+        // query all slug.
+        $mlSlug = $translateModel::findFirst([
+            'conditions' => 'foreign_id = :id: AND lang = :lang: AND key = "slug"',
+            'bind'       => [
+                'id'     => $slugRecord->foreign_id,
+                'lang'   => $lang
+            ],
+        ]);
+
+        return $mlSlug->value;
```
Comment (sergeyklay) [inline app/modules/Application/Mvc/Helper/LangSwitcher.php]: what happens if `$mlSlug` is null?

File: app/modules/Page/Model/Page.php [hunk 10]
```diff
@@ -65,9 +66,16 @@ public function validation()
 
     public static function findCachedBySlug($slug)
     {
-        $query = "slug = '$slug'";
+        $queryML = "key = 'slug' and value = '$slug'";
+        $key = HOST_HASH . md5("Translate\PageTranslate::findFirst($queryML)");
```
Comment (sergeyklay) [inline app/modules/Page/Model/Page.php]: Just use `__METHOD__`

File: app/modules/Page/Model/Page.php [hunk 11]
```diff
@@ -65,9 +66,16 @@ public function validation()
...
+        $slugRecord = Translate\PageTranslate::findFirst([$queryML, 'cache' => array('key' => $key, 'lifetime' => 60)]);
+
         $key = HOST_HASH . md5("Page::findFirst($query)");
```
Comment (sergeyklay) [inline app/modules/Page/Model/Page.php]: Just use `__METHOD__`

File: app/modules/Application/Mvc/Helper/LangSwitcher.php (hunk 7 above)
Comment (sergeyklay) [inline app/modules/Application/Mvc/Helper/LangSwitcher.php]: Also note: the `getMLSlug` will be used in public `render`. So once you made it private the `render` can't be extended in child class.

**Comment by oleksandr-torosh** [comment]
What is the main reason of this Pull Request?

**Comment by oleksandr-torosh** [comment]
Sorry, I can't merge it

**Pull request closed without merging**